File: app/src/main/java/de/kuschku/util/backports/Optional.java
/*
 * QuasselDroid - Quassel client for Android
 * Copyright (C) 2016 Janne Koschinski
 * Copyright (C) 2016 Ken Børge Viktil
 * Copyright (C) 2016 Magnus Fjell
 * Copyright (C) 2016 Martin Sandsmark <martin.sandsmark@kde.org>
 *
 * This program is free software: you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by the Free
 * Software Foundation, either version 3 of the License, or (at your option)
 * any later version, or under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either version 2.1 of
 * the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License and the
 * GNU Lesser General Public License along with this program.  If not, see
 * <http://www.gnu.org/licenses/>.
 */

package de.kuschku.util.backports;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import com.google.common.base.Function;
import com.google.common.base.Predicate;
import com.google.common.base.Supplier;

public interface Optional<T> {
    @NonNull
    Optional<T> filter(@NonNull Predicate<? super T> predicate);

    @Nullable
    <U> Optional<U> flatMap(@NonNull Function<? super T, Optional<U>> mapper);

    @NonNull
    <U> Optional<U> map(@NonNull Function<? super T, U> mapper);

    @Nullable
    T get();

    void ifPresent(@NonNull Consumer<? super T> consumer);

    boolean isPresent();

    @NonNull
    T orElse(@NonNull T other);

    @Nullable
    T orElseGet(@NonNull Supplier<? extends T> other);

    @NonNull
    <X extends Throwable> T orElseThrow(@NonNull Supplier<? extends X> exceptionSupplier) throws X;
}

File: lib/src/main/java/de/kuschku/libquassel/util/Optional.kt
/*
 * Quasseldroid - Quassel client for Android
 *
 * Copyright (c) 2019 Janne Mareike Koschinski
 * Copyright (c) 2019 The Quassel Project
 *
 * This program is free software: you can redistribute it and/or modify it
 * under the terms of the GNU General Public License version 3 as published
 * by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */

package de.kuschku.libquassel.util

import java.io.Serializable

interface Optional<T : Any> : Serializable {
  fun get(): T
  fun isPresent(): Boolean
  fun ifPresent(consumer: (T) -> Unit)
  fun filter(predicate: (T) -> Boolean): Optional<T>
  fun <U : Any> map(mapper: (T) -> U): Optional<U>
  fun <U : Any> flatMap(mapper: (T) -> Optional<U>): Optional<U>
  fun orElse(other: T): T
  fun orNull(): T?
  fun <X : Throwable> orElseThrow(supplier: () -> X): T
  override fun equals(other: Any?): Boolean
  override fun hashCode(): Int
  override fun toString(): String

  private class Present<T : Any>(private val value: T) : Optional<T> {
    override fun get() = value
    override fun isPresent() = true
    override fun ifPresent(consumer: (T) -> Unit) = consumer(value)
    override fun filter(predicate: (T) -> Boolean) = if (predicate(value)) this else empty<T>()
    override fun <U : Any> map(mapper: (T) -> U) = ofNullable(mapper(value))
    override fun <U : Any> flatMap(mapper: (T) -> Optional<U>) = mapper(value)
    override fun orElse(other: T) = value
    override fun orNull(): T? = value
    override fun <X : Throwable> orElseThrow(supplier: () -> X) = value
    override fun equals(other: Any?) = (other as? Present<*>)?.value == value
    override fun hashCode() = value.hashCode()
    override fun toString() = "Optional[$value]"
  }

  private class Absent<T : Any> : Optional<T> {
    override fun get() = throw NoSuchElementException("No value present")
    override fun isPresent() = false
    override fun ifPresent(consumer: (T) -> Unit) = Unit
    override fun filter(predicate: (T) -> Boolean) = this
    override fun <U : Any> map(mapper: (T) -> U) = empty<U>()
    override fun <U : Any> flatMap(mapper: (T) -> Optional<U>) = empty<U>()
    override fun orElse(other: T) = other
    override fun orNull(): T? = null
    override fun <X : Throwable> orElseThrow(supplier: () -> X) = throw supplier()
    override fun equals(other: Any?) = other === this
    override fun hashCode() = 0
    override fun toString() = "Optional.empty"
  }

  companion object {
    private val absent = Absent<Any>()

    @Suppress("UNCHECKED_CAST")
    fun <T : Any> empty(): Optional<T> = absent as Absent<T>

    fun <T : Any> of(value: T): Optional<T> = Present(value)
    fun <T : Any> ofNullable(value: T?): Optional<T> = value?.let(::Present) ?: empty()
  }
}
